src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,583 @@
+    case PredicateCondition::LessThan:
+    case PredicateCondition::LessThanEquals:
+    case PredicateCondition::GreaterThan:
+    case PredicateCondition::GreaterThanEquals:
+      break;
+    default:
+      return nullptr;
+  }
+
+  // We can currently only pull equals predicates above aggregate nodes (by grouping by the column that the predicate
+  // compares with). The other predicate types could be supported but would require more sophisticated reformulations.
+  if (is_below_aggregate && predicate_condition != PredicateCondition::Equals) {
+    return nullptr;
+  }
+
+  // Check that one side of the expression is a correlated parameter and the other a column expression of the LQP below
+  // the predicate node (required for turning it into a join predicate). Also order the left/right operands by the
+  // subtrees they originate from.
+  const auto& binary_predicate_expression = std::static_pointer_cast<BinaryPredicateExpression>(predicate_expression);
+  const auto& left_side = binary_predicate_expression->left_operand();
+  const auto& right_side = binary_predicate_expression->right_operand();
+  auto ordered_predicate = binary_predicate_expression->deep_copy();
+  ParameterID parameter_id;
+  std::shared_ptr<AbstractExpression> right_operand;
+  if (left_side->type == ExpressionType::CorrelatedParameter) {
+    parameter_id = std::static_pointer_cast<CorrelatedParameterExpression>(left_side)->parameter_id;
+    right_operand = right_side;
+  } else if (right_side->type == ExpressionType::CorrelatedParameter) {
+    predicate_condition = flip_predicate_condition(predicate_condition);
+    parameter_id = std::static_pointer_cast<CorrelatedParameterExpression>(right_side)->parameter_id;
+    right_operand = left_side;
+  } else {
+    return nullptr;
+  }
+
+  // We can only use predicates in joins where both operands are columns
+  if (!predicate_node->find_column_id(*right_operand)) {
+    return nullptr;
+  }
+
+  // Is the parameter one we are concerned with? This catches correlated parameters of outer subqueries and
+  // placeholders in prepared statements.
+  auto expression_it = parameter_mapping.find(parameter_id);
+  if (expression_it == parameter_mapping.end()) {
+    return nullptr;
+  }
+
+  auto left_operand = expression_it->second;
+  return std::make_shared<BinaryPredicateExpression>(predicate_condition, left_operand, right_operand);
+}
+
+/**
+ * Copy an aggregate node and adapt it to group by all required columns.
+ */
+std::shared_ptr<AggregateNode> adapt_aggregate_node(
+    const AggregateNode& node, const std::vector<std::shared_ptr<AbstractExpression>>& required_column_expressions) {
+  std::vector<std::shared_ptr<AbstractExpression>> group_by_expressions(
+      node.node_expressions.cbegin(), node.node_expressions.cbegin() + node.aggregate_expressions_begin_idx);
+  ExpressionUnorderedSet original_group_by_expressions(group_by_expressions.cbegin(), group_by_expressions.cend());
```

Comment:
Just wanted to mention again that I prefer the "auto to stick"-style, see [here why](https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/).
```suggestion
  auto original_group_by_expressions = ExpressionUnorderedSet(group_by_expressions.cbegin(), group_by_expressions.cend());
```